tycho-gpg-plugin/src/main/java/org/apache/maven/plugin/gpg/ProxySignerWithPublicKeyAccess.java

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Red Hat Inc. and others.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.apache.maven.plugin.gpg;
+
+import java.io.File;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.codehaus.plexus.util.Os;
+import org.codehaus.plexus.util.StringUtils;
+import org.codehaus.plexus.util.cli.CommandLineException;
+import org.codehaus.plexus.util.cli.CommandLineUtils;
+import org.codehaus.plexus.util.cli.CommandLineUtils.StringStreamConsumer;
+import org.codehaus.plexus.util.cli.Commandline;
+
+public class ProxySignerWithPublicKeyAccess extends AbstractGpgSigner {
+
+    private AbstractGpgSigner delegate;
+
+    public ProxySignerWithPublicKeyAccess(AbstractGpgSigner newSigner) {
+        this.delegate = newSigner;
+    }
+
+    @Override
+    protected void generateSignatureForFile(File file, File signature) throws MojoExecutionException {
+        delegate.generateSignatureForFile(file, signature);
+    }
+
+    public String getPublicKeys() throws MojoExecutionException {
+        Commandline cmd = new Commandline();
+
+//        if ( StringUtils.isNotEmpty( executable ) ) {
+//            cmd.setExecutable( executable );
+//        } else {
+        cmd.setExecutable("gpg" + (Os.isFamily(Os.FAMILY_WINDOWS) ? ".exe" : ""));
```

Comment:
This would require pgp installed right? Wouldn't it be better to use an of the java implementations of PGP (like jgit for git) instead if rely on the OS? https://www.openpgp.org/software/developer/

Reply:
This is using the same startegy as the regular maven-gpg-plugin This brings consistency, although it would be nice to be able to leverage more of the maven-pgp-plugin as API.
One benefit of using the CLI is that it also reuse other gpg settings user might have tailored on their OS. If using boucycastle, it means we have to deal with building the keyring and so on. I have spent some time with bouncycastle on p2 earlier, and it's something I'd like to avoid as much as possible when we can: it's difficult to write and maintain, and it requires to be extra careful to not leak some keys and things like that. Security is a can of worm, so when it comes built-in as a binary we can expect users who need it have it, it's better to use it IMO.

Reply:
> although it would be nice to be able to leverage more of the maven-pgp-plugin as API

What would be needed here? can we request this from the maven-pgp-plugin directly?

> One benefit of using the CLI is that it also reuse other gpg settings user might have tailored on their OS

Don't know if gpg has similar default locations like maven (.m2) or git (.git/config) that could be adopted? Especially as we are only using the signing part, and the "secret key" is already to be configured I wonder what would could be inherited here?

Calling some system-executable really is something I would prevent as it also opens a "can of worms", just think about I write some malicious code named gpg.exe especially at windows ...

Reply:
> What would be needed here? can we request this from the maven-pgp-plugin directly?

Basically everything that we have in the org.apache.maven.plugin.pgp package here would need to become more public. It would indeed be a good thing to request it to the maven-gpg-plugin.

> Calling some system-executable really is something I would prevent as it also opens a "can of worms", just think about I write some malicious code named gpg.exe especially at windows ...

Except that if this happens, then it's an issue of the environment, not of Tycho. So Tycho would still be correct invoking gpg.exe and the issue would happen with every other consumer. That reduce the responsibility load on Tycho.
In any case, we should just mimic what's done in the official maven-pgp-plugin instead of trying to do things differently, The requirement on the binary is minimal: anyone caring about pgp has a gpg binary installed.